In [108]:
import pandas as pd
import numpy as np

In [109]:
df_housing = pd.read_csv("../data/eda.csv")
df_housing.head()

,id,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,...,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15,date,price
0,7129300520,3.0,1.00,1180.0,5650.0,1.0,NaN,0.0,3,7,...,0.0,1955,0.0,98178,47.5112,-122.257,1340.0,5650.0,2014-10-13,221900.0
1,6414100192,3.0,2.25,2570.0,7242.0,2.0,0.0,0.0,3,7,...,400.0,1951,19910.0,98125,47.7210,-122.319,1690.0,7639.0,2014-12-09,538000.0
2,5631500400,2.0,1.00,770.0,10000.0,1.0,0.0,0.0,3,6,...,0.0,1933,NaN,98028,47.7379,-122.233,2720.0,8062.0,2015-02-25,180000.0
3,2487200875,4.0,3.00,1960.0,5000.0,1.0,0.0,0.0,5,7,...,910.0,1965,0.0,98136,47.5208,-122.393,1360.0,5000.0,2014-12-09,604000.0
4,1954400510,3.0,2.00,1680.0,8080.0,1.0,0.0,0.0,3,8,...,0.0,1987,0.0,98074,47.6168,-122.045,1800.0,7503.0,2015-02-18,510000.0


In [110]:
df_housing.info()
df_housing.describe()

<class 'pandas.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   bedrooms       21597 non-null  float64
 2   bathrooms      21597 non-null  float64
 3   sqft_living    21597 non-null  float64
 4   sqft_lot       21597 non-null  float64
 5   floors         21597 non-null  float64
 6   waterfront     19206 non-null  float64
 7   view           21534 non-null  float64
 8   condition      21597 non-null  int64  
 9   grade          21597 non-null  int64  
 10  sqft_above     21597 non-null  float64
 11  sqft_basement  21145 non-null  float64
 12  yr_built       21597 non-null  int64  
 13  yr_renovated   17749 non-null  float64
 14  zipcode        21597 non-null  int64  
 15  lat            21597 non-null  float64
 16  long           21597 non-null  float64
 17  sqft_living15  21597 non-null  float64
 18  sqft_lot15     21

,id,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15,price
count,2.159700e+04,21597.000000,21597.000000,21597.000000,2.159700e+04,21597.000000,19206.000000,21534.000000,21597.000000,21597.000000,21597.000000,21145.000000,21597.000000,17749.000000,21597.000000,21597.000000,21597.000000,21597.000000,21597.000000,2.159700e+04
mean,4.580474e+09,3.373200,2.115826,2080.321850,1.509941e+04,1.494096,0.007602,0.233863,3.409825,7.657915,1788.596842,291.857224,1970.999676,836.650516,98077.951845,47.560093,-122.213983,1986.620318,12758.283512,5.402966e+05
std,2.876736e+09,0.926299,0.768984,918.106125,4.141264e+04,0.539683,0.086858,0.765686,0.650546,1.173200,827.759761,442.490863,29.375234,4000.110554,53.513072,0.138552,0.140724,685.230472,27274.441950,3.673681e+05
min,1.000102e+06,1.000000,0.500000,370.000000,5.200000e+02,1.000000,0.000000,0.000000,1.000000,3.000000,370.000000,0.000000,1900.000000,0.000000,98001.000000,47.155900,-122.519000,399.000000,651.000000,7.800000e+04
25%,2.123049e+09,3.000000,1.750000,1430.000000,5.040000e+03,1.000000,0.000000,0.000000,3.000000,7.000000,1190.000000,0.000000,1951.000000,0.000000,98033.000000,47.471100,-122.328000,1490.000000,5100.000000,3.220000e+05
50%,3.904930e+09,3.000000,2.250000,1910.000000,7.618000e+03,1.500000,0.000000,0.000000,3.000000,7.000000,1560.000000,0.000000,1975.000000,0.000000,98065.000000,47.571800,-122.231000,1840.000000,7620.000000,4.500000e+05
75%,7.308900e+09,4.000000,2.500000,2550.000000,1.068500e+04,2.000000,0.000000,0.000000,4.000000,8.000000,2210.000000,560.000000,1997.000000,0.000000,98118.000000,47.678000,-122.125000,2360.000000,10083.000000,6.450000e+05
max,9.900000e+09,33.000000,8.000000,13540.000000,1.651359e+06,3.500000,1.000000,4.000000,5.000000,13.000000,9410.000000,4820.000000,2015.000000,20150.000000,98199.000000,47.777600,-121.315000,6210.000000,871200.000000,7.700000e+06


In [111]:
df_housing.duplicated().value_counts()

False    21597
Name: count, dtype: int64

Data Cleaning Step 1: Convert column date to datetime format

In [112]:
df_housing["date"] = pd.to_datetime(df_housing["date"], format="%Y-%m-%d")

Note on `bathrooms` decimals:
Decimal values in `bathrooms` etc are not data errors — they follow a standard US real estate convention where each fixture (toilet, sink, shower, tub) counts as 0.25. For example, 0.5 = half bath, 0.75 = three-quarter bath, and 1.75 = one full bathroom + one three-quarter bathroom. We will leave these values unchanged.

Data Cleaning Step 2: Column yr_renovated contains 3858 (17.8%) NaN values. which we are interpreting as no renovation as no date was mentioned for renovation, therefore replacing with 0 and converting to INT data type to get rid of the decimals. Additionally this column contains an extra 0 at the end of all years which needs to be removed by dividing by 10.

In [113]:
df_housing.info()

<class 'pandas.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   id             21597 non-null  int64         
 1   bedrooms       21597 non-null  float64       
 2   bathrooms      21597 non-null  float64       
 3   sqft_living    21597 non-null  float64       
 4   sqft_lot       21597 non-null  float64       
 5   floors         21597 non-null  float64       
 6   waterfront     19206 non-null  float64       
 7   view           21534 non-null  float64       
 8   condition      21597 non-null  int64         
 9   grade          21597 non-null  int64         
 10  sqft_above     21597 non-null  float64       
 11  sqft_basement  21145 non-null  float64       
 12  yr_built       21597 non-null  int64         
 13  yr_renovated   17749 non-null  float64       
 14  zipcode        21597 non-null  int64         
 15  lat            21597 non-null 

In [114]:
df_housing["yr_renovated"].isna().sum()

np.int64(3848)

In [115]:
df_housing["yr_renovated"].info()

<class 'pandas.Series'>
RangeIndex: 21597 entries, 0 to 21596
Series name: yr_renovated
Non-Null Count  Dtype  
--------------  -----  
17749 non-null  float64
dtypes: float64(1)
memory usage: 168.9 KB


In [116]:
df_housing.yr_renovated.isna().sum() / df_housing.shape[0] * 100

np.float64(17.81728943834792)

In [117]:
df_housing["yr_renovated"] = df_housing["yr_renovated"].fillna(0)
df_housing["yr_renovated"].isna().sum()

np.int64(0)

In [118]:
df_housing["yr_renovated"]

0            0.0
1        19910.0
2            0.0
3            0.0
4            0.0
          ...   
21592        0.0
21593        0.0
21594        0.0
21595        0.0
21596        0.0
Name: yr_renovated, Length: 21597, dtype: float64

In [119]:
df_housing = df_housing.astype({"yr_renovated": int})
df_housing["yr_renovated"]

0            0
1        19910
2            0
3            0
4            0
         ...  
21592        0
21593        0
21594        0
21595        0
21596        0
Name: yr_renovated, Length: 21597, dtype: int64

In [120]:
df_housing["yr_renovated"].unique()

array([    0, 19910, 20020, 20100, 19920, 20130, 19940, 19780, 20050,
       20030, 19840, 19540, 20140, 20110, 19830, 19450, 19900, 19880,
       19770, 19810, 19950, 20000, 19990, 19980, 19700, 19890, 20040,
       19860, 20070, 19870, 20060, 19850, 20010, 19800, 19710, 19790,
       19970, 19500, 19690, 19480, 20090, 20150, 19740, 20080, 19680,
       20120, 19630, 19510, 19620, 19530, 19930, 19960, 19550, 19820,
       19560, 19400, 19760, 19460, 19750, 19640, 19730, 19570, 19590,
       19600, 19670, 19650, 19340, 19720, 19440, 19580])

In [121]:
mask = df_housing["yr_renovated"] > 2100
df_housing.loc[mask, "yr_renovated"] = df_housing.loc[mask, "yr_renovated"] / 10
df_housing["yr_renovated"].unique()

array([   0, 1991, 2002, 2010, 1992, 2013, 1994, 1978, 2005, 2003, 1984,
       1954, 2014, 2011, 1983, 1945, 1990, 1988, 1977, 1981, 1995, 2000,
       1999, 1998, 1970, 1989, 2004, 1986, 2007, 1987, 2006, 1985, 2001,
       1980, 1971, 1979, 1997, 1950, 1969, 1948, 2009, 2015, 1974, 2008,
       1968, 2012, 1963, 1951, 1962, 1953, 1993, 1996, 1955, 1982, 1956,
       1940, 1976, 1946, 1975, 1964, 1973, 1957, 1959, 1960, 1967, 1965,
       1934, 1972, 1944, 1958])

In [122]:
df_housing["yr_renovated"].dtypes

dtype('int64')

In [123]:
import missingno as msno

In [124]:
df_housing.isna().sum()

id                  0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2391
view               63
condition           0
grade               0
sqft_above          0
sqft_basement     452
yr_built            0
yr_renovated        0
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
date                0
price               0
dtype: int64

Data Cleaning Step 3: Fill NA values for waterfront, Column waterfront = 1 is extremely rare, under 1% of the total house from this data set, therefore we will replace missing values with 0 and convert to Int.

In [126]:
df_housing["waterfront"].value_counts(normalize=True)

waterfront
0.0    0.992398
1.0    0.007602
Name: proportion, dtype: float64

In [127]:
df_housing["waterfront"] = df_housing["waterfront"].fillna(0)
df_housing["waterfront"].isna().sum()

np.int64(0)

In [129]:
df_housing.info()

<class 'pandas.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   id             21597 non-null  int64         
 1   bedrooms       21597 non-null  float64       
 2   bathrooms      21597 non-null  float64       
 3   sqft_living    21597 non-null  float64       
 4   sqft_lot       21597 non-null  float64       
 5   floors         21597 non-null  float64       
 6   waterfront     21597 non-null  float64       
 7   view           21534 non-null  float64       
 8   condition      21597 non-null  int64         
 9   grade          21597 non-null  int64         
 10  sqft_above     21597 non-null  float64       
 11  sqft_basement  21145 non-null  float64       
 12  yr_built       21597 non-null  int64         
 13  yr_renovated   21597 non-null  int64         
 14  zipcode        21597 non-null  int64         
 15  lat            21597 non-null 

In [130]:
df_housing["waterfront"].unique()

array([0., 1.])

In [131]:
df_housing["waterfront"] = df_housing["waterfront"].astype(int)

In [133]:
df_housing["view"].value_counts(normalize=True)

view
0.0    0.901923
2.0    0.044441
3.0    0.023591
1.0    0.015325
4.0    0.014721
Name: proportion, dtype: float64

Data Clean Step 4: Fill NA for view. Same logic as waterfront because only 63 values from view are missing and the data shows around 90% of the houses have 0 for a view. For simplicity we also convert this to int

In [134]:
df_housing["view"] = df_housing["view"].fillna(0)
df_housing["view"] = df_housing["view"].astype(int)
df_housing["view"].value_counts(normalize=True)

view
0    0.902209
2    0.044312
3    0.023522
1    0.015280
4    0.014678
Name: proportion, dtype: float64

In [136]:
df_housing["view"].dtypes

dtype('int64')

Data Clean Step 5: Replace missing values for the sqft_basement column. First we verify if there is a relationship between the different sqft values such as sqft_basement = sqft_living - sqft_above

In [137]:
df_housing.isna().sum()

id                 0
bedrooms           0
bathrooms          0
sqft_living        0
sqft_lot           0
floors             0
waterfront         0
view               0
condition          0
grade              0
sqft_above         0
sqft_basement    452
yr_built           0
yr_renovated       0
zipcode            0
lat                0
long               0
sqft_living15      0
sqft_lot15         0
date               0
price              0
dtype: int64

In [139]:
known = df_housing.dropna(subset = ["sqft_basement"])
(known["sqft_living"] - known["sqft_above"] == known["sqft_basement"]).all()

np.True_

In [141]:
maskh = df_housing["sqft_basement"].isna()
df_housing.loc[maskh, "sqft_basement"] = (
    df_housing.loc[maskh, "sqft_living"] - df_housing.loc[maskh, "sqft_above"]
)

In [142]:
df_housing.isna().sum()

id               0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
date             0
price            0
dtype: int64

In [143]:
df_housing.head()

,id,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,...,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15,date,price
0,7129300520,3.0,1.00,1180.0,5650.0,1.0,0,0,3,7,...,0.0,1955,0,98178,47.5112,-122.257,1340.0,5650.0,2014-10-13,221900.0
1,6414100192,3.0,2.25,2570.0,7242.0,2.0,0,0,3,7,...,400.0,1951,1991,98125,47.7210,-122.319,1690.0,7639.0,2014-12-09,538000.0
2,5631500400,2.0,1.00,770.0,10000.0,1.0,0,0,3,6,...,0.0,1933,0,98028,47.7379,-122.233,2720.0,8062.0,2015-02-25,180000.0
3,2487200875,4.0,3.00,1960.0,5000.0,1.0,0,0,5,7,...,910.0,1965,0,98136,47.5208,-122.393,1360.0,5000.0,2014-12-09,604000.0
4,1954400510,3.0,2.00,1680.0,8080.0,1.0,0,0,3,8,...,0.0,1987,0,98074,47.6168,-122.045,1800.0,7503.0,2015-02-18,510000.0


In [ ]:
df_housing.isna().all()

id               False
bedrooms         False
bathrooms        False
sqft_living      False
sqft_lot         False
floors           False
waterfront       False
view             False
condition        False
grade            False
sqft_above       False
sqft_basement    False
yr_built         False
yr_renovated     False
zipcode          False
lat              False
long             False
sqft_living15    False
sqft_lot15       False
date             False
price            False
dtype: bool

In [147]:
df_housing.dtypes

id                        int64
bedrooms                float64
bathrooms               float64
sqft_living             float64
sqft_lot                float64
floors                  float64
waterfront                int64
view                      int64
condition                 int64
grade                     int64
sqft_above              float64
sqft_basement           float64
yr_built                  int64
yr_renovated              int64
zipcode                   int64
lat                     float64
long                    float64
sqft_living15           float64
sqft_lot15              float64
date             datetime64[us]
price                   float64
dtype: object

In [150]:
df_housing.duplicated().sum()      # any exact duplicate rows?   # confirm total NAs across the whole df is 0


np.int64(0)

In [151]:
df_housing.isna().sum().sum()

np.int64(0)